ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,279 @@
+	zonesCount := uint32(len(t.zones))
+	optimalTokenOwnership = (optimalTokenOwnership / zonesCount) * zonesCount
+	return optimalTokenOwnership
+}
+
+// GenerateTokens generates at most tokensCount unique tokens, none of which clash with the given takenTokens.
+// Generated tokens are sorted. The number of returned tokens might be less than requested in the following cases:
+//   - if tokensCount is higher than the maximal configured number of tokens of the
+//     underlying instance (t.cfg.tokensPerInstance).
+//   - if among t.cfg.tokensPerInstance tokens reserved for the underlying instance
+//     there is less than tokenCount tokens not already present in takenTokens.
+func (t *SpreadMinimizingTokenGenerator) GenerateTokens(tokensCount int, takenTokens []uint32) Tokens {
+	used := make(map[uint32]bool, len(takenTokens))
+	for _, v := range takenTokens {
+		used[v] = true
+	}
+
+	allTokens := t.generateAllTokens()
+	tokens := make(Tokens, 0, tokensCount)
+
+	// allTokens is a sorted slice of tokens for instance t.cfg.InstanceID in zone t.cfg.zone
+	// We filter out tokens from takenTokens, if any, and return at most tokensCount tokens.
+	for i := 0; i < len(allTokens) && len(tokens) < tokensCount; i++ {
+		token := allTokens[i]
+		if used[token] {
+			continue
+		}
+		tokens = append(tokens, token)
+	}
+	return tokens
+}
+
+// generateAllTokens generates the configured number of tokens (t.cfg.tokensPerInstance) for the underlying instance.
+// Generated tokens are sorted. Calls to this method will always return the same set of tokens.
+func (t *SpreadMinimizingTokenGenerator) generateAllTokens() Tokens {
+	firstInstanceTokens := t.generateFirstInstanceTokens()
+
+	if t.instanceID == 0 {
+		return firstInstanceTokens
+	}
+
+	tokensCount := t.cfg.TokensPerInstance
+	// tokensQueues is a slice of priority queues. Slice indexes correspond
+	// to the ids of instances, while priority queues represent the tokens
+	// of the corresponding instance, ordered from highest to lowest ownership.
+	tokensQueues := make([]ownershipPriorityQueue[ringToken], t.instanceID)
+
+	// Create and initialize priority queue of tokens for the first instance
+	tokensQueue := newPriorityQueue[ringToken](tokensCount)
+	prev := len(firstInstanceTokens) - 1
+	firstInstanceOwnership := 0.0
+	for tk, token := range firstInstanceTokens {
+		tokenOwnership := float64(getTokenDistance(firstInstanceTokens[prev], token))
+		firstInstanceOwnership += tokenOwnership
+		tokensQueue.Add(newRingTokenOwnershipInfo(token, firstInstanceTokens[prev]))
+		prev = tk
+	}
+	heap.Init(&tokensQueue)
+	tokensQueues[0] = tokensQueue
+
+	// instanceQueue is a priority queue of instances such that instances with higher ownership have a higher priority
+	instanceQueue := newPriorityQueue[ringInstance](t.instanceID)
+	instanceQueue.Add(newRingInstanceOwnershipInfo(0, firstInstanceOwnership))
+	heap.Init(&instanceQueue)
```

Comment:
Note that `heap.Init` call is useless for a single element, but if we really want to use heap, we can just `heap.Push(&instanceQueue, newRingInstanceOwnershipInfo(0, firstInstanceOwnership))`

Reply:
Thank you for this hint. I thought it was mandatory to call `heap.Init` before doing any other operation.

Reply:
`heap.Init()` will just ensure that it's a valid heap, but a slice of one element is already a max heap :D (and a min heap!)